Monolane generator

wjwwood opened the pull request:
This pull request adds the experimental monolane generator (both the module for it and adds it to the `run_generator.py` script).

It also changes slightly how the builders are imported in the `run_generator.py` script because my editor cannot do local variable checking when `from X import *` is used. We can exclude that commit if you guys want, but I think it is best to avoid that kind of import in general.

---

The monolane generator is able to produce a monolane `.yaml` file which can be converted to an `.obj` file using the `yaml_to_obj` executable from drake. This is the command I use to convert the `.yaml` into the `.obj` file:

```
../drake/bazel-bin/drake/automotive/maliput/utility/yaml_to_obj -yaml_file ./generated_worlds/city.monolane.yaml -obj_file city -obj_dir ./generated_worlds
```

Obviously you'd have to have already built drake (in this case I'm using bazel to do so rather than cmake) and also replace the path to the executable according to where you built it.

However, there is still an issue with the geometry and so the default version of this executable (`yaml_to_obj`) will fail without disabling some of the error checking. I used this patch:

```diff
diff --git a/drake/automotive/maliput/monolane/builder.cc b/drake/automotive/maliput/monolane/builder.cc
index 2bc60fb..c0afd02 100644
--- a/drake/automotive/maliput/monolane/builder.cc
+++ b/drake/automotive/maliput/monolane/builder.cc
@@ -306,7 +306,7 @@ std::unique_ptr<const api::RoadGeometry> Builder::Build(
   for (const auto& s : failures) {
     drake::log()->error(s);
   }
-  DRAKE_DEMAND(failures.size() == 0);
+  // DRAKE_DEMAND(failures.size() == 0);
 
   return std::move(road_geometry);
 }
```

The generation still works despite the errors which look something like this:

```
[2017-02-03 17:54:38.445] [console] [error] Lane l:trunk_88-5-trunk_88-6[start] orientation is off by 0.0574029.
```

To solve this we will need to address some of the geometric issues with the road layouts. What RNDF needs doesn't exactly line up with what monolane expects, but I have stopped short of implementing this because I believe these kinds of geometry calculations would be useful outside of the monolane generator and probably should be integrated into terminus directly. I will create an issue about this in more detail after this is merged.

Also, if the road segment is very short (common in the OSM output) then drake will not convert because the arrow it wants to draw on the road is longer than the road. This isn't fatal so I disabled that as well with this patch:

```diff
diff --git a/drake/automotive/maliput/utility/generate_obj.cc b/drake/automotive/maliput/utility/generate_obj.cc
index 86e117c..86d7f08 100644
--- a/drake/automotive/maliput/utility/generate_obj.cc
+++ b/drake/automotive/maliput/utility/generate_obj.cc
@@ -408,7 +408,7 @@ void StripeLaneBounds(GeoMesh* mesh, const api::Lane* lane,
 void DrawLaneArrow(GeoMesh* mesh, const api::Lane* lane, double grid_unit,
                    double s_offset, double s_size, double h_offset) {
   DRAKE_DEMAND(s_offset >= 0.);
-  DRAKE_DEMAND((s_offset + s_size) <= lane->length());
+  // DRAKE_DEMAND((s_offset + s_size) <= lane->length());
   const double kRelativeWidth = 0.8;
 
   const api::RBounds rb0 = lane->lane_bounds(s_offset);

```

I believe this should be addressed upstream in drake at some point. I'll open an issue there after this is merged.

----

I think use `meshlab` to view the `.obj` file. Here are some examples of what is generated:

`SimpleCityBuilder`:

![screenshot from 2017-02-03 17-52-06](https://cloud.githubusercontent.com/assets/100427/22614626/8973fee8-ea39-11e6-82c4-66ce847fba74.png)

OSM at OSRF:

![screenshot from 2017-02-03 17-53-21](https://cloud.githubusercontent.com/assets/100427/22614636/b0c10040-ea39-11e6-9145-632187200cbf.png)

`ProceduralCityBuilder`:

![screenshot from 2017-02-03 17-56-04](https://cloud.githubusercontent.com/assets/100427/22614657/11fb92d0-ea3a-11e6-872a-63d61ea97bfd.png)

----

Other than the "orientation is off by" errors (which make the output of this generator still not very useful since it requires drake patches to parse it), there are two more main issues that need to be addressed. Both are related to intersections and I'll make new issues for those when this is merged, but I'll briefly describe both of them.

The first is that monolane typically creates arcs between road non-co-linear connections (both at intersections and normal nodes) to have continuous geometry (this is related to the orientation error). For example, this is what one of their hand crafted monolane files looks like after obj generation:

![screenshot from 2017-02-03 18-01-41](https://cloud.githubusercontent.com/assets/100427/22614712/dad6aa28-ea3a-11e6-9349-15be618b396d.png)

But in contrast this is the center intersection for the simple city based on this generator:

![screenshot from 2017-02-03 18-02-57](https://cloud.githubusercontent.com/assets/100427/22614721/038040a6-ea3b-11e6-8db9-728b1e7208f2.png)

I think this might need to be addressed more generally in terminus since it would probably be useful to have non-disjoint road geometry for other generators too (maybe gazebo?).

The second is how the connections in monolane are created by this generator. It creates many unrelated connections between points in space, rather than chaining off of other existing connections. This might have adverse effects on how the monolane file is used programmatically, but I'm not certain about that.

CC @nkoenig

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,203 @@
+import yaml
+
+from .city_visitor import CityVisitor
+from models.intersection_node import IntersectionNode
```
`IntersectionNode` was removed and now we have `RoadIntersectionNode` for roads and `LaneIntersectionNode` (but based on the code I guess you want `RoadIntersectionNode`)

wjwwood replied:
adb6385

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,203 @@
+        self.intersection_index = 0
+        self.intersection_index_by_node = {}
+
+    def node_ident(self, street, node):
```
It would be great to decouple the id->node mapping methods and instance variables (AFAIK only `point_name_by_street_node_dict`) to a separate object, like the `RNDFIdMapper`. We found a similar need for OpenDrive and created the (quite dummy) `OpenDriveIdMapper` so it would be great to have a third requirement for this to see if we can generalize it to a single class.

wjwwood replied:
I have to admit, I found the fact that a `CityVisitor` ran another `CityVisitor` to be very confusing and was one of the reasons I avoided copying the rndf generator directly. It seemed unnecessarily complicated. I don't mind refactoring the the storage and methods for calculating the identity to another class, but I'd rather not make it a `CityVisitor` like `RNDFIdMapper` unless you really want it that way.

It's not clear to me if this can be generalized or not, as each format may have different rules for what may go into a name, but you might be able to generalize the interface. I'll look and see where there is overlap with what I am doing and what is in the interface for the other `*IdMapper` classes.

basicNew replied:
The usage of a visitor for that case came from RNDF, as it maps ids in a hierarchical fashion (`segment_id.lane_id.waypoint_id`) so using the visitor was a natural way of mapping that recursive path. Agree 100% though that, if that is not a requirement, there is no need to implement it as a visitor (and I'm not 100% positive that it can be generalized either :)). I just saw that a similar requirement (doing a first pass to generate an id->object lookup table) was present in two generators before, so I thought it would be a good think to separate them and see if we can find at least (as you mention) a common interface. With that first step we can see if we can later reuse some behavior.

wjwwood replied:
I refactored the monolane generator to produce the point names (same as ids in RNDF) in a separate class which is also a `CityVisitor`. I then changed it so the connections were made as streets were processes rather than in `end_city()`.

7d176f2

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,203 @@
+    def start_trunk(self, trunk):
+        self.process_street(trunk)
+
+    def end_trunk(self, trunk):
```
I don't think  you need to redefine all these `start_*`/`end_*` as they have empty implementation on the superclass

wjwwood replied:
Yeah, that was for convenience of adding print statements while I was debugging, I'll remove the extra ones: a0d6276

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,203 @@
+    def end_building(self, building):
+        pass
+
+    def to_string(self, stream=None, Dumper=yaml.Dumper, **kwds):
```
Even though you don't subclass from `FileGenerator`, could you please implement this in terms of the `generate` / `write_to` public interface? This will make it much easier to refactor later.

wjwwood replied:
Yeah, I can do that.

I, again, avoided this feature of the `RNDFGenerator` because it didn't make sense to me to conflate the writing to files with the generator, at least in the monolane case, but I could see how it might be important if the generator needed to make multiple files. I'll refactor it to either inherit from `FileGenerator` or at least adhere to the interface you described.

wjwwood replied:
d28dcef

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,203 @@
+from models.intersection_node import IntersectionNode
+
+
+class MonolaneGenerator(CityVisitor):
```
It would be great to have a couple of simple tests to understand what we expect this class to generate. You can use the same samples in `RNDFGeneratorTest` if you want (now that I'm thinking, it would be great to use the same examples, we can later factor them out and use them as the standard tests for all the generators).

wjwwood replied:
Sure thing, I'll look at those tests and see if I can easily adapt them to the monolane generator.

wjwwood replied:
Ok, I refactored the `RNDFGenerator` test cases into a separate module in 6616c51 and used those to implement similar tests for the `MonolaneGenerator` in dcaa31a.

basicNew commented:
Regarding the  `import *`, it is perfect to change it to your proposal.

basicNew commented:
Regarding the `orientation is off by` I think (and this is mostly a hunch based on a quick look to the code) the problem is that the `yaml_to_obj` is using spherical coordinates while we are working in the 3D cartesian space. I created a small example to reproduce the issue, which is:

```
# -*- yaml -*-
---
# distances are meters; angles are degrees.
maliput_monolane_builder:
  id: city_1
  lane_bounds: [-2, 2]
  driveable_bounds: [-4, 4]
  position_precision: 0.01
  orientation_precision: 0.5
  points:
    street_2-2:
      xypoint: [100.0, 100.0, 90.0]
      zpoint: [0.0, 0, 0, 0]
    street_2-3:
      xypoint: [100.0, 200.0, 90.0]
      zpoint: [0.0, 0, 0, 0]
    street_6-1:
      xypoint: [-2.0, 200.0, 0.0]
      zpoint: [0.0, 0, 0, 0]
    street_6-2:
      xypoint: [100.0, 200.0, 0.0]
      zpoint: [0.0, 0, 0, 0]
  connections:
    street_2-2-street_2-3: {start: points.street_2-2, length: 100.0, explicit_end: points.street_2-3}
    street_6-1-street_6-2: {start: points.street_6-1, length: 102.0, explicit_end: points.street_6-2}
  groups: {}
```

Which seems a valid one in a three dimensional cartesian coordinate system. However this yields:

```
[2017-02-06 18:49:34.205] [console] [info] Loading road geometry.
[2017-02-06 18:49:34.208] [console] [error] Lane l:street_2-2-street_2-3[end] orientation is off by 2.22144.
```

So maybe the first thing would be to make sure that we are producing the file in the right coordinate system.

basicNew commented:
Regarding the intersections, it definitely makes sense, as we are just using the geometry nodes of the polylines. As we briefly discussed before, the points providing the before / after offset are the waypoints, which are now moved to the lanes instead of the roads. So what do you think about having an `as_single_lane_street()` method in `Road`? This would return a road with the same geometry as the one defined, but with a single lane at offset `0`, hence mimicking the road geometry. And then you can ask the lane for the waypoints instead of the nodes, which will give you the middle points you need. Let me know if this makes sense an I can push that change.

wjwwood commented:
I added the dependency on `PyYAML` (monolane is output in yaml) to the `setup.py` too.

wjwwood commented:
(also rebased on master)

wjwwood commented:
Looks like CI is passing in f4d4d92, I've rebased and addressed the original set of comments.

@basicNew I'd like to propose merging this in its current state (unless there are other comments since the new commits), after which I would immediately open issues about it relating to the other three points about the road geometry and intersections. That way I can bring in monolane people into the discussion using more narrowly scoped issues.

wjwwood commented:
> Regarding the `orientation is off by` I think (and this is mostly a hunch based on a quick look to the code) the problem is that the `yaml_to_obj` is using spherical coordinates while we are working in the 3D cartesian space.

That may be the case, but I have some toy monolane files which leads me to believe that it is not simple the heading of the points that is the issue (and therefore fixing their coordinate system shouldn't make a difference). I'd like to hash this out in a separate issue on which I can bring in some monolane users to get their ideas too.

wjwwood commented:
> As we briefly discussed before, the points providing the before / after offset are the waypoints, which are now moved to the lanes instead of the roads. So what do you think about having an `as_single_lane_street()` method in `Road`?

I think this might be useful, but it also (seems to me) to be conflating the idea of nodes being the geometric anchor for segments consisting of multiple lanes, where lanes would have have waypoints to define them which are distributed laterally within the road, and the other idea of using additional waypoints near the intersections to set ourselves up to make more geometrically natural connections between roads (or lanes) at an intersection. In both cases waypoints might be the answer, but for monolane we'd also need to calculate the arcs between these new "intersection waypoints" and we'd have to be more specific about which lanes can connect to which other lanes through an intersection.

Again, I'd like to open more narrow issues for this topic.

basicNew commented on terminus/generators/monolane_generator.py:
```diff
@@ -0,0 +1,103 @@
+from __future__ import print_function
```
Just to confirm, is this required or was a leftover from debugging phase?

basicNew commented on terminus/generators/monolane_id_mapper.py:
```diff
@@ -0,0 +1,120 @@
+    def road_node_id(self, road, node):
+        """Return a hashable unique identifier for a road node."""
+        if road.name is None:
+            print(road)
```
Not sure if this is supposed to happen or not. Can you replace this for either an exception if it should break or a `logger.debug` if it should remain for debugging purposes but not fail?

wjwwood replied:
Nope, that was just debugging. Removed it and the future imports as well.

basicNew commented on terminus/run_generator.py:
```diff
@@ -1,12 +1,18 @@
 #!/usr/bin/python
 
+from __future__ import print_function
```
Same as before, just checking we still need this.

basicNew commented on terminus/run_generator.py:
```diff
@@ -60,8 +66,15 @@
+    ProceduralCityBuilder,
+    SimpleCityBuilder
+]
+assert(arguments.builder in [x.__name__ for x in builders_list])
```
+1, thanks!

wjwwood commented:
Should be ready now.

wjwwood commented:
Thanks, I'll start opening the other issues.

maddog-tri commented:
There are a few key things to understand about `maliput` first:
 * Maliput models the geometry of a road network as a surface which is C1-continuous --- i.e., it must have continuous derivatives everywhere.  (One reason is so that normals/tangents/orientations are well-defined everywhere.  Another is so that the _surface_ is continuous/contiguous, even off the lane centerline.)
 * `monolane` is pretty much the simplest implementation ("backend") of the maliput API that allows for the creation of interesting road networks.  It has two geometric primitives, line segments and circular arcs, which is the minimum to make a non-straight road that observes C1-continuity.
 * the `monolane::Builder` API, and the associated "yaml format" is designed to make it easy to hook up lines and arcs while maintaining continuity --- to make it easy to construct non-trivial road networks by hand. As you've discovered, the syntax can be abused.

The error you are seeing is the `maliput::api` invariant checker complaining about the discontinuities:  the end of one lane is not parallel to the end of the lane it joins to.  And, as you've seen, if you hack the code to produce an OBJ anyway, there are overlapping swaths and wedges of missing pavement instead of continuous surfaces around corners.

So, yes, all bends in the road need an arc (or two, or more) somewhere.  Intersections basically need arc-connections to all possible roads off the straight-through path --- this provides topological connectivity (the choices available when driving through the intersection). The arcs should have some radius that is appropriate to the generated road --- e.g., the radius must be at least the size of the driveable-bounds of road.  (Oh, and all connections that spatially overlap at an intersection should be in the same "group".)

For procedurally-generated roads that are basically "straight edges connected at point-vertices" the work is just synthesizing some equally arbitrary rounded-corners at the vertices (which are the intersections).  This is fairly simple in the sense that, if you are making up a road network out of thin air, then you are free to connect the dots any way you want (as long as they are continuous! :).

For roads from the real world (OSM data) that are not arbitrary and not straight and are defined by many intermediate waypoints, it will probably require some more analysis to do some interpolation between waypoints, basically doing a "piecewise-arc" traversal between intersections.  It all depends on how close you want to get to the 'ground-truth' (though, I'm guessing that OSM doesn't set a very high bar for precision to begin with).

wjwwood commented:
> So, yes, all bends in the road need an arc (or two, or more) somewhere.

Great so that basically confirms my conclusion:

> > Long story short, I think we need to use arcs between segments in the road and at intersections to avoid the orientation errors from the `yaml_to_obj` program in drake.
>
> -- https://github.com/ekumenlabs/terminus/issues/185

---

> (Oh, and all connections that spatially overlap at an intersection should be in the same "group".)

Ok, that's another thing that's missing then. I asked about this on slack at some point and the consensus of the people there was that it was unused at the moment. I just created an issue to track that here: https://github.com/ekumenlabs/terminus/issues/188

---

For both cases, procedurally generated roads and data gathered from OSM, I think it makes sense to generate these "connecting geometries" (arcs between bends in the road) in terminus and utilize them in the monolane generator, because they might also be useful in other generators as well.

Thanks for the input @maddog-tri. The only question I still have is related to using connection ends or explicit points when defining new connections, see: https://github.com/ekumenlabs/terminus/issues/187 If you could have a look at that one, I would really appreciate it.
